2.x: Add scheduler creation factories

ZacSweers opened the pull request:
Resolves #4993

This is a pretty vanilla copy from RxJava 1's implementation. Note that I had to tune NewThread scheduler to not be a singleton to support this.

We had talked about borrowing from project reactor's APIs for different overloads, let me know if you think we should add more fine-grained controls through these.

ZacSweers committed:
Add scheduler creation factories

Resolves #4993

This is a pretty vanilla copy from RxJava 1's implementation. Note that I had to tune NewThread scheduler to not be a singleton to support this.

We had talked about borrowing from project reactor's APIs for different overloads, let me know if you think we should add more fine-grained controls through these.

ZacSweers commented:
One other thing I'm worried about - is `Schedulers` the best place for this? RxJava 1 had them in its `RxJavaHooks` class. My concern is that having these alongside the normal `io()`/`computation()`/etc factories will make them look overly inviting for use and accidental abuse.

akarnokd commented on src/main/java/io/reactivex/internal/schedulers/IoScheduler.java:
```diff
@@ -56,6 +54,9 @@
 
         EVICTOR_THREAD_FACTORY = new RxThreadFactory(EVICTOR_THREAD_NAME_PREFIX, priority);
+
+        NONE = new CachedWorkerPool(0, null, WORKER_THREAD_FACTORY);
```
Any particular reason why this was moved to the end?

ZacSweers replied:
It requires WORKER_THREAD_FACTORY to be initialized now since it's a parameter

akarnokd replied:
Right.

akarnokd commented on src/main/java/io/reactivex/internal/schedulers/SingleScheduler.java:
```diff
@@ -47,11 +49,20 @@
+     *                      system properties for configuring new thread creation. Cannot be null.
+     */
+    public SingleScheduler(ThreadFactory threadFactory) {
+        this.threadFactory = ObjectHelper.requireNonNull(threadFactory, "threadFactory was null");
```
Either this is unnecessary or you forgot to do this with the other new constructors.

ZacSweers replied:
Oops, meant to remove it from all these constructors and just do it in the factory methods, accidentally missed this one

ZacSweers replied:
e5c4a9e

akarnokd commented on src/main/java/io/reactivex/schedulers/Schedulers.java:
```diff
@@ -180,6 +181,90 @@ public static Scheduler from(Executor executor) {
     /**
+     * Create an instance of the default {@link Scheduler} used for {@link Schedulers#computation()}.
+     * @return the created Scheduler instance
+     */
```
please add `@since 2.0.5 - experimental` as well for every new method.

ZacSweers replied:
Will do

ZacSweers replied:
262f669

akarnokd commented on src/main/java/io/reactivex/schedulers/Schedulers.java:
```diff
@@ -180,6 +181,90 @@ public static Scheduler from(Executor executor) {
+     * @return the created Scheduler instance
+     */
+    @Experimental
+    public static Scheduler newComputation() {
```
I'm not sure we'd want to allow non-parameterized instances to be created.

ZacSweers replied:
Could you expand on that a bit?

akarnokd replied:
What is the use case for creating yet another computation/io/newThread scheduler with the default settings? I thought you wanted to override the `ThreadFactory`.

akarnokd commented on src/main/java/io/reactivex/schedulers/Schedulers.java:
```diff
@@ -180,6 +181,90 @@ public static Scheduler from(Executor executor) {
+     */
+    @Experimental
+    public static Scheduler newComputation(ThreadFactory threadFactory) {
+        return new ComputationScheduler(ObjectHelper.requireNonNull(threadFactory, "threadFactory == null"));
```
The usual NPE message is `threadFactory is null`.

ZacSweers replied:
Gotcha, will update

ZacSweers replied:
681b977

akarnokd commented on src/main/java/io/reactivex/schedulers/Schedulers.java:
```diff
@@ -180,6 +181,90 @@ public static Scheduler from(Executor executor) {
+    }
+
+    /**
+     * Create an instance of the default {@link Scheduler} used for {@link Schedulers#computation()}
```
These need some rewording emphasizing that the there is always a new Scheduler instance, unlike with the non-new method call.

akarnokd commented:
> One other thing I'm worried about - is Schedulers the best place for this? RxJava 1 had them in its RxJavaHooks class. My concern is that having these alongside the normal io()/computation()/etc factories will make them look overly inviting for use and accidental abuse.

Another problem is that people will try to override the scheduler-init with these method calls for `RxJavaPlugins` but referencing `Schedulers` will init the default schedulers regardless.

Indeed, `RxJavaPlugins` sounds like a better place for the methods, but keep only the `ThreadFactory` variants.

ZacSweers commented:
Heading to bed right now but will update the PR in the morning with requested changes. Also let me know if there's anywhere I should put tests, as I wasn't sure what the right place would be.

ZacSweers commented:
> Indeed, RxJavaPlugins sounds like a better place for the methods, but keep only the ThreadFactory variants.

Sounds good to me, will do that in my updates tomorrow

akarnokd commented:
Yes, tests that verify the custom thread factory actually worked by checking a custom thread name for each case.

ZacSweers committed:
Add `@since` info

ZacSweers committed:
Change failure string to "is null"

ZacSweers committed:
Move to RxJavaPlugins

ZacSweers committed:
Remove no-arg overloads

ZacSweers committed:
Rename to make it clearer about creation

Added scheduler because we're not in Scheduler anymore. Changed to "create" because "newNewThread" was weird

ZacSweers committed:
Add tests (WIP)

ZacSweers committed:
Remove unnecessary nullcheck

akarnokd commented on src/test/java/io/reactivex/plugins/RxJavaPluginsTest.java:
```diff
@@ -1890,4 +1888,106 @@ public void accept(final Throwable throwable) throws Exception {
+        ThreadFactory factory = new ThreadFactory() {
+            @Override
+            public Thread newThread(Runnable r) {
+                return new Thread("Test");
```
I think this doesn't allow executing tasks because you ignore `r`.

ZacSweers commented:
> Indeed, RxJavaPlugins sounds like a better place for the methods

8b4d461

> but keep only the ThreadFactory variants.

8009333

Also tweaked the naming a bit. `newNewThread` was a little weird (went with `create`), and added `Scheduler` suffix since we're not in the `Schedulers` class anymore.

> Yes, tests that verify the custom thread factory actually worked by checking a custom thread name for each case.

I tried setting this up matching some of the cdl-based approaches in the plugin tests but wan't able to get it working (just hangs). Pushed what I had in a1029b4, any insight?

akarnokd commented on src/test/java/io/reactivex/plugins/RxJavaPluginsTest.java:
```diff
@@ -1890,4 +1888,106 @@ public void accept(final Throwable throwable) throws Exception {
+                return "Test".equals(thread.getName());
+            }
+        });
+    }
```
You have to shut down the scheduler (in a finally block) to avoid leaking threads.

akarnokd commented on src/main/java/io/reactivex/plugins/RxJavaPlugins.java:
```diff
@@ -927,6 +929,58 @@ public static Completable onAssembly(Completable source) {
     /**
+     * Create an instance of the default {@link Scheduler} used for {@link Schedulers#computation()}
+     * except using {@code threadFactory} for thread creation.
+     * @param threadFactory thread factory to use for creating worker threads. Note that this takes precedence over any
```
Please add notes to all of these methods something like this:

> Note that the returned `Scheduler` must be shut down manually if the `ThreadFactory` doesn't  create a daemon thread, otherwise the JVM may not quit.

akarnokd replied:
The "Note that this takes precedence..." may be misleading. The number of threads is still based on the system configuration. I'd remove this sentence entirely

ZacSweers commented:
Heading on vacation for a couple weeks, but let me know what you think of getting the tests to work. I should have some time here and there to update the PR

akarnokd commented:
I've already told you: use the runnable in the factory method on the Thread constructor.

ZacSweers commented:
> use the runnable in the factory method on the Thread constructor.

Aren't I just testing the test ThreadFactory implementations then, and not that the schedulers are hooked up properly to power the scheduler?

akarnokd commented:
https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.html

https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/schedulers/RxThreadFactory.java#L66

JakeWharton commented on src/test/java/io/reactivex/plugins/RxJavaPluginsTest.java:
```diff
@@ -1890,4 +1888,106 @@ public void accept(final Throwable throwable) throws Exception {
+
+    private static void verifyThread(Worker w, Predicate<Thread> threadPredicate) {
+        try {
+            try {
```
remove double try

ZacSweers replied:
90ae664

JakeWharton commented on src/test/java/io/reactivex/plugins/RxJavaPluginsTest.java:
```diff
@@ -1890,4 +1888,106 @@ public void accept(final Throwable throwable) throws Exception {
+    private static void verifyThread(Worker w, Predicate<Thread> threadPredicate) {
+        try {
+            try {
+                final AtomicReference<Thread> value = new AtomicReference<>();
```
Maybe change to `AtomicBoolean` and just run the predicate inside the worker?

JakeWharton replied:
Then you don't even need to supply the `Thread` to the `Predicate` actually. It can verify whatever it wants.

JakeWharton commented:
What's the problem with the tests? They seem to be verifying that the supplied factory was used.

JakeWharton commented:
Oh you didn't pass the supplied `Runnable` when calling `new Thread()` inside the factory.

ZacSweers committed:
Remove double try

ZacSweers committed:
Fix tests, make them more robust with integration flow

ZacSweers commented:
> Oh you didn't pass the supplied Runnable when calling new Thread() inside the factory.

Ah! This is what I was missing, thanks for pointing that out. I've updated tests (which should now all be passing), and also made them a bit more robust (full integration with schedulers) to simulate a more real world use case.

ZacSweers commented:
Failing test looks like a flake...

ZacSweers commented:
Wait no it's not. Repro'd locally, will fix

ZacSweers committed:
Shut down custom schedulers when done

ZacSweers commented:
Fixed, I think. I added manual shutdowns of schedulers after they're done. Only IO had this lifecycle issue over tests, but I did the shutdown in all to be safe.

ZacSweers commented:
Ok this time it seems actually flaky, as the same commit passed in a different travis job against my fork - https://travis-ci.org/hzsweers/RxJava/builds/195074173

akarnokd committed:
Merge branch '2.x' into schedulerFactories

ZacSweers commented:
Thanks!
